File: tokenizer.py
# tokenizer.py

import re

def tokenize(code):
    token_specification = [
        ('TEMPLATE', r'`[^`]*`'),
        ('STRING',   r"'[^']*'"),
        ('NUMBER',   r'\d+'),
        ('BOOLEAN',  r'\btrue\b|\bfalse\b'),
        ('TYPE',     r'\bstring\b|\bint\b|\bbool\b'),
        ('CONST',    r'\bconst\b'),
        ('AS',       r'\bas\b'),
        ('EMIT',     r'\bemit\b'),
        ('IDENT',    r'[a-zA-Z_][a-zA-Z0-9_]*'),
        ('EQUALS',   r'='),
        ('PLUS',     r'\+'),
        ('MINUS',    r'-'),
        ('STAR',     r'\*'),
        ('SLASH',    r'/'),
        ('SEMICOLON',r';'),
        ('SKIP',     r'[ \t\n]+'),
        ('MISMATCH', r'.'),
    ]
    tok_regex = '|'.join(f'(?P<{name}>{pattern})' for name, pattern in token_specification)
    get_token = re.compile(tok_regex).match
    pos = 0
    mo = get_token(code, pos)
    tokens = []
    while mo:
        kind = mo.lastgroup
        value = mo.group()
        if kind == 'SKIP':
            pass
        elif kind == 'MISMATCH':
            raise RuntimeError(f'Unexpected character: {value!r}')
        else:
            tokens.append((kind, value))
        pos = mo.end()
        mo = get_token(code, pos)
    tokens.append(('EOF', ''))
    return tokens

File: tokenizer.js
// tokenizer.js

export function tokenize(code) {
    const tokenSpecification = [
        ['TEMPLATE', /`[^`]*`/y],
        ['STRING', /'[^']*'/y],
        ['NUMBER', /\d+/y],
        ['BOOLEAN', /\btrue\b|\bfalse\b/y],
        ['TYPE', /\bstring\b|\bint\b|\bbool\b/y],
        ['CONST', /\bconst\b/y],
        ['AS', /\bas\b/y],
        ['EMIT', /\bemit\b/y],
        ['IDENT', /[a-zA-Z_][a-zA-Z0-9_]*/y],
        ['EQUALS', /=/y],
        ['PLUS', /\+/y],
        ['MINUS', /-/y],
        ['STAR', /\*/y],
        ['SLASH', /\//y],
        ['SEMICOLON', /;/y],
        ['SKIP', /[ \t\r\n]+/y],
        ['MISMATCH', /./y]
    ];

    const tokens = [];
    let pos = 0;

    while (pos < code.length) {
        let match = null;
        for (const [name, regex] of tokenSpecification) {
            regex.lastIndex = pos;
            match = regex.exec(code);
            if (match && match.index === pos) {
                const value = match[0];
                if (name === 'SKIP') {
                    // skip whitespace
                } else if (name === 'MISMATCH') {
                    throw new Error(`Unexpected character: '${value}' at position ${pos}`);
                } else {
                    tokens.push([name, value]);
                }
                pos += value.length;
                break;
            }
        }
        if (!match) {
            throw new Error(`Unexpected token at position ${pos}`);
        }
    }
    tokens.push(['EOF', '']);
    return tokens;
}
